prow/plugins/dco/dco.go

```diff
@@ -0,0 +1,334 @@
+	}
+	l.Infof("Found %d commits in PR", len(allCommits))
+
+	var shasMissingDCO []string
+	for _, commit := range allCommits {
+		if !testRe.MatchString(*commit.Commit.Message) {
+			shasMissingDCO = append(shasMissingDCO, *commit.SHA)
+		}
+	}
+
+	l.Infof("All commits in PR have DCO signoff: %t", len(shasMissingDCO) == 0)
+	return shasMissingDCO, nil
+}
+
+// checkExistingStatus will retrieve the current status of the DCO context for
+// the provided SHA.
+func checkExistingStatus(gc gitHubClient, l *logrus.Entry, org, repo, sha string) (string, error) {
+	statuses, err := gc.ListStatuses(org, repo, sha)
+	if err != nil {
+		return "", fmt.Errorf("error listing pull request statuses: %v", err)
+	}
+
+	existingStatus := ""
+	for _, status := range statuses {
+		if status.Context != dcoContextName {
+			continue
+		}
+		existingStatus = status.State
+		break
+	}
+	l.Infof("Existing DCO status context status is %q", existingStatus)
+	return existingStatus, nil
+}
+
+// checkExistingLabels will check the provided PR for the dco sign off labels,
+// returning bool's indicating whether the 'yes' and the 'no' label are present.
+func checkExistingLabels(gc gitHubClient, l *logrus.Entry, org, repo string, number int) (hasYesLabel, hasNoLabel bool, err error) {
+	labels, err := gc.GetIssueLabels(org, repo, number)
+	if err != nil {
+		return false, false, fmt.Errorf("error getting pull request labels: %v", err)
+	}
+
+	for _, l := range labels {
+		if l.Name == dcoYesLabel {
+			hasYesLabel = true
+		}
+		if l.Name == dcoNoLabel {
+			hasNoLabel = true
+		}
+	}
+
+	return hasYesLabel, hasNoLabel, nil
+}
+
+// takeAction will take appropriate action on the pull request according to its
+// current state.
+func takeAction(gc gitHubClient, cp commentPruner, l *logrus.Entry, org, repo string, pr github.PullRequest, commitsMissingDCO []string, existingStatus string, hasYesLabel, hasNoLabel, addComment bool) error {
+	targetURL := fmt.Sprintf("https://github.com/%s/%s/blob/master/CONTRIBUTING.md", org, repo)
+	botName, err := gc.BotName()
+	if err != nil {
+		return fmt.Errorf("failed to get bot name: %v", err)
+	}
+
+	signedOff := len(commitsMissingDCO) == 0
+
+	// handle the 'all commits signed off' case by adding appropriate labels
+	// TODO: clean-up old comments?
+	if signedOff {
+		if hasNoLabel {
+			l.Infof("Removing %q label", dcoNoLabel)
+			// remove 'dco-signoff: no' label
+			if err := gc.RemoveLabel(org, repo, pr.Number, dcoNoLabel); err != nil {
+				return fmt.Errorf("error removing label: %v", err)
+			}
+		}
+		if !hasYesLabel {
+			l.Infof("Adding %q label", dcoYesLabel)
+			// add 'dco-signoff: yes' label
+			if err := gc.AddLabel(org, repo, pr.Number, dcoYesLabel); err != nil {
+				return fmt.Errorf("error adding label: %v", err)
+			}
+		}
+		if existingStatus != github.StatusSuccess {
+			l.Infof("Setting DCO status context to succeeded")
+			if err := gc.CreateStatus(org, repo, pr.Head.SHA, github.Status{
```

Comment:
Do we want to label every commit that is valid? For instance:

 - create PR with valid commits A B C
 - force-push to remove commit C

I guess right now we will just re-calculate everything, does it matter?

Reply:
I'm not too sure what we'd like to do here. I asked [here](https://github.com/kubernetes/test-infra/pull/9197#issuecomment-416953651) and expanded a bit on the behaviour of the cncf-cla bot (which marks *all commits*).

Marking all commits seemed like a bad idea though wrt API token usage 🙄 (I could just create a 5k commit PR and then Prow dies)